loki/transform/transform_lift_contained_subroutines.py

```diff
@@ -0,0 +1,212 @@
+            if varname.lower() in varnames_in_import:
+                return impo.clone()
+        raise Exception(f"variable '{varname}' was not found from the specified import objects.")
+
+    routine = routine.clone()
+
+    # Get all inner routines and resolve all associate statements in them.
+    # This is strictly not needed, but is done to simplify the logic in later steps.
+    inner_routines = []
+    for r in routine.contains.body:
+        if isinstance(r, Subroutine):
+            newinner = r.clone()
+            resolve_associates(newinner)
+            inner_routines.append(newinner)
+    assert len(inner_routines) > 0
+
+    # Get all calls.
+    all_calls = FindNodes(CallStatement).visit(routine.body)
+    call_transmap = {}
+
+    # Construct a list of routines returned from this function.
+    routines = []
+
+    outer_decls = FindNodes(VariableDeclaration).visit(routine.spec)
+
+    for inner in inner_routines:
+        # Find all local variables (names). 
+        # This includes:
+        # 1. Declared non-array variables in spec.
+        # 2. Array-variables, EXCLUDING any variables that only exist in array dimensions.
+        # NOTE: Such dimension variables will be in group 1, if they are actually defined.
+        # 3. Imported variables. NOTE: Here collecting also imports that correspond to functions,
+        # but they are removed later. 
+        ldecls = FindNodes(VariableDeclaration).visit(inner.spec)
+        limports_vars = FindVariables().visit(FindNodes(Import).visit(inner.spec))
+        lvars = list(reduce(lambda x, y: x + y, (decl.symbols for decl in ldecls))) 
+        lvars = set(var.clone() for var in lvars + list(limports_vars))
+        lvar_names = [var.name.lower() for var in lvars]
+
+        # Find all variables (names), including also:
+        # 1. Declared non-array and array variables in spec.
+        # 2. Variables in body.
+        # 3. Variable kinds.
+        # 4. Array dimension variables.
+        all_vars_body = FindVariables().visit(inner.body)
+        all_vars_spec = FindVariables().visit(inner.spec)
+        all_vars = set(var.clone() for var in all_vars_body.union(all_vars_spec))
+        kinds = set()
+        for var in all_vars:
+            if var.type.kind:
+                kinds.add(var.type.kind.clone())
+        all_vars = all_vars.union(kinds)
+
+        # Get the names of all global variables from the perspective of `inner`. 
+        gvar_names = set()
+        for var in all_vars:
+            # The next while loop is in case `var` is a member of a derived type. 
+            # For example for variable named `a%b%c`, 
+            # the global variable we want is `a`, not `a%b%c` itself.
+            varname = var.name.lower()
+            vartmp = var.clone()
+            while vartmp.parent is not None:
+                varname = vartmp.parent.name.lower()
+                vartmp = vartmp.parent
```

Comment:
Perhaps you meant something different with this, because that change would break the code.
The idea here is to have `varname` point to the "root" of a derived type (in case `var` IS a derived type).
In other words, if `var` was something like `a%b%c`, the while loop walks the tree backwards to get `varname` = "a". Cannot just look at a single parent, therefore. I suspect this might be refactored away with your next comments, however..

Reply:
In fact, this is exactly what I understood and proposed. Note the difference between `vartmp.parent` (which will be None if it isn't a derived type member) and `vartmp.parents[0]`, which yields all parents and take the outermost - which is exactly the chasing you implemented.

Reply:
Apologies, I must have read your code in haste. (somehow I missed that it was `parents[0]`, not `parent`, as you explained) I see your point now. Still, I did run the tests with your suggestion, and recall getting an error (maybe it was `IndexError` for `parents` being empty in some special case), or then it was something unrelated.

However, note that this bit got removed in my latest refactoring, and there is no need to "walk to the parent" anymore, since it appears I can just drop all variables that have a parent.

Reply:
Right, sorry, I had been making my way through the comments only and hadn't realized this bit had gone entirely. Ignore my ramblings then ;-) I'll try to get through the rest of it today.

Reply:
It is good you mentioned it, I had not noticed there is also the `parents` field.